### PR TITLE
Review

### PR DESCRIPTION
**Pull request by Javiercerna**

### CODE REVIEW

File: bagOfWords.py [hunk 1]
```diff
@@ -0,0 +1,61 @@
+import re
+import csv
+from bs4 import BeautifulSoup
+from nltk.tokenize import word_tokenize
+from nltk.corpus import stopwords
+from sklearn.feature_extraction.text import CountVectorizer
```
Comment (Javiercerna) [inline bagOfWords.py:6]:
```suggestion
import csv
import re

from bs4 import BeautifulSoup
from nltk.corpus import stopwords
from nltk.tokenize import word_tokenize
from sklearn.feature_extraction.text import CountVectorizer
```

For our Python style, we work with PEP-8, so your code would probably be refactored as this. It's nothing serious, just to let you know how we work.

File: bagOfWords.py [hunk 2]
```diff
@@ -0,0 +1,61 @@
...
+
+def removeNonLetter(sentence):
+	filteredSentence = []
+	sentence = word_tokenize(sentence)
+	for word in sentence:
+		if word.isalpha():
+			filteredSentence.append(word)
```
Comment (Javiercerna) [inline bagOfWords.py:13]:
```suggestion
    for word in sentence:
        if word.isalpha():
            filteredSentence.append(word)
```
Be careful with your indentations :slightly_smiling_face: Again, nothing serious, but it's part of our conventions

File: bagOfWords.py [hunk 3]
```diff
@@ -0,0 +1,61 @@
...
+	return sentence
+
+#main program
+fp = open('reviews.csv','r')
+line = fp.readline()
+dataset = []
+for i in range(0,1600):
+	line = fp.readline()
+	dataset.append(preProcessing(line))
+fp.close()
```
Comment (Javiercerna) [inline bagOfWords.py:28]:
```suggestion
dataset = []
with open('reviews.csv', 'r') as fp:
    for i in range(0,1600):
        line = fp.readline()
        dataset.append(preProcessing(line))
```
Here I would recommend you to use the `with` context manager to handle opening and closing files. I'm not sure if you have a particular reason to use `range(0, 1600)` to read lines, but unless there is some particular meaning to that number, I would read line by line like this:

```
for line in fp:
    dataset.append(preProcessing(line))
```

https://docs.python.org/3/tutorial/inputoutput.html#methods-of-file-objects

File: bagOfWords.py [hunk 4]
```diff
@@ -0,0 +1,61 @@
...
+
+text = dataset
```
Comment (Javiercerna) [inline bagOfWords.py:30]: Why are you copying this to another variable? :sweat_smile:

File: bagOfWords.py [hunk 5]
```diff
@@ -0,0 +1,61 @@
...
+	f.write(",")
+	for word in uniqueWords:
+		f.write(word+",")
+	f.write("\n")
+	for i in range(len(text)):
+		f.write(text[i]+",")
+		for j in countvector[i]:
+			f.write(str(j)+",")
+		f.write("\n")
+f.close
```
Comment (Javiercerna) [inline bagOfWords.py:61]:
```suggestion
f.close()
```

Typo :slightly_smiling_face:

File: bagOfWords.py [hunk 6]
```diff
@@ -0,0 +1,61 @@
...
+
+# encode document
+vector = vectorizer.transform(text)
+
+# summarize encoded vector
+print(vector.shape)
+countvector = vector.toarray()
+
+#write to file
+f = open("countVector.csv", 'w')
```
Comment (Javiercerna) [inline bagOfWords.py:50]: Same as before, use the context manager `with`

File: bigramProbability.py [hunk 7]
```diff
@@ -0,0 +1,44 @@
...
+	all_words.append(w.lower())
+
+new_words = []
+for w in all_words:
+	if w.isalpha():
+		new_words.append(w)
+
+stopWords = set(stopwords.words("english"))
+character = {'u'}
+tokensFiltered = []
```
Comment (Javiercerna) [inline bigramProbability.py:25]:
```suggestion
tokens_filtered = []
```

Again, some PEP-8 comments. Even ignoring that, our focus is on consistency. So since you have already defined other variables such as `all_words` or `new_words`, you should stick to that convention.

And, of course, we reserve camelCase to methods

File: bigramProbability.py [hunk 8]
```diff
@@ -0,0 +1,44 @@
...
+for w in movie_reviews.words():
+	all_words.append(w.lower())
+
+new_words = []
+for w in all_words:
+	if w.isalpha():
+		new_words.append(w)
+
+stopWords = set(stopwords.words("english"))
+character = {'u'}
```
Comment (Javiercerna) [inline bigramProbability.py:24]:
```suggestion
```
I think you don't use this variable anywhere, so it doesn't hurt to remove it to clean your code

File: feature_extraction.py [hunk 9]
```diff
@@ -0,0 +1,55 @@
...
+    return sentence
+
+def remove_nonletter(sentence):
+    filtered_sentence = []
+    for w in sentence:
+        if w.isalpha:
+            filtered_sentence.append(w)
+    return filtered_sentence
+
+def stopword_removal(sentence):
```
Comment (Javiercerna) [inline feature_extraction.py:23]:
```suggestion
def stopWordRemoval(sentence):
```

File: feature_extraction.py [hunk 10]
```diff
@@ -0,0 +1,55 @@
...
+            filtered_sentence.append(word)
+    return filtered_sentence
+
+def getUnigrams(token):
+	unigramsDictionary.update(Counter(token))
+	return
+
+def getBigrams(token):
+    bigramsDictionary.update(Counter(token))
+    return
```
Comment (Javiercerna) [inline feature_extraction.py:37]:
```suggestion
def updateUnigrams(token):
	unigramsDictionary.update(Counter(token))

def updateBigrams(token):
    bigramsDictionary.update(Counter(token)) 
```
Here I would suggest you to not use `get`, since you are actually not returning anything. Instead, you are updating, which I think it's pretty self-explanatory. By the way, you can take advantage of Python implicit `None` returns in this case.

File: feature_extraction.py [hunk 11]
```diff
@@ -0,0 +1,55 @@
...
+line = fp.readline()
+for line in fp:
+    token = pre_processing(line)
+    getUnigrams(token)
+    getBigrams(token)
+fp.close()
+
+for key,val in bigramsDictionary.items():
+	p = "p("+key[1]+"/"+key[0]+")"
+	bigramsProbability[p] = float(val)/unigramsDictonary[key[0]]
```
Comment (Javiercerna) [inline feature_extraction.py:55]: Here and in other scripts I would suggest you to use the notation:

```
if __name__ == '__main__':
    # Your code
```
It makes it much easier to export your methods to other files if they need to. If you are not sure about the reason of my suggestion, please check here: https://stackoverflow.com/questions/419163/what-does-if-name-main-do

File: feature_vector.py [hunk 12]
```diff
@@ -0,0 +1,75 @@
...
+            counts[gram] = 1
+        else:
+            counts[grams]+=1
+    with open('bigramcount.csv','w+') as f:
+        w = csv.writer(f)
+        w.writerows(counts.items())
+    f.close()
+    return
+
+def NumericalFeatureVector(processed_text):
```
Comment (Javiercerna) [inline feature_vector.py:52]: This one is a red flag for me!

```suggestion
def numericalFeatureVector(processed_text):
```
By your definition it looks like you were working with a class, and that could be confusing at first sight. Something as simple as using the correct convention can solve the entire issue.

Of course, it feels a bit weird that you have a method called `numericalFeatureVector`. What does it do? Maybe it should be called `computeNumericalFeatureVector` instead, but it's up to you.

File: feature_vector.py [hunk 13]
```diff
@@ -0,0 +1,75 @@
...
+            vectorizer.fit(w)
+            vector = vectorizer.transform(w)
+            numerical_feature_vectors = vector.toarray()
+
+def pre_processing(data):
+    sentence = to_lower(data)
+    sentence = nonletter_removal(sentence)
+    filtered_sentence = remove_stopwords(sentence)
+    #getNgrams(filtered_sentence,2)
+    NumericalFeatureVector(filtered_sentence)
```
Comment (Javiercerna) [inline feature_vector.py:65]: What is this method doing? Inside you assign values to a `numerical_feature_vectors` variable, but you never use it and this method is not returning any values either :confused:

File: sentiment_mining.py [hunk 14]
```diff
@@ -0,0 +1,36 @@
+import nltk
+import random
+from nltk.corpus import movie_reviews
+
+documents = [(list(movie_reviews.words(fileid)), category)
+             for category in movie_reviews.categories()
+             for fileid in movie_reviews.fileids(category)]
```
Comment (Javiercerna) [inline sentiment_mining.py:7]: Nice use of list comprehensions!

File: sentiment_mining.py [hunk 15]
```diff
@@ -0,0 +1,36 @@
...
+
+random.shuffle(documents)
+
+all_words = []
+
+for w in movie_reviews.words():
+    all_words.append(w.lower())
```
Comment (Javiercerna) [inline sentiment_mining.py:14]: You could also use list comprehensions here :smile: 

```suggestion
all_words = [w.lower() for w in movie_reviews.words()]
```

File: sentiment_mining.py [hunk 16]
```diff
@@ -0,0 +1,36 @@
...
+
+all_words = nltk.FreqDist(all_words)
+
+word_features = list(all_words.keys())[:1000]
```
Comment (Javiercerna) [inline sentiment_mining.py:18]: Why 1000? Maybe you could define a constant called `LIMIT_WORDS` or something similar that you use to control this and make it more explicit.

File: sentiment_mining.py [hunk 17]
```diff
@@ -0,0 +1,36 @@
...
+
+def find_features(document):
```
Comment (Javiercerna) [inline sentiment_mining.py:20]: I would move this method to the top of the document, so you can keep a more structured layout:

```
# Method definitions (if you have any)
# Class definitions (if you have any)
# if __name__ == '__main__' code
```

File: sentiment_mining.py [hunk 18]
```diff
@@ -0,0 +1,36 @@
...
+    words = set(document)
+    features = {}
+    for w in word_features:
+        features[w] = (w in words)
+    return features
+
+featuresets = [(find_features(rev), category) for (rev, category) in documents]
+
+training_set = featuresets[:700]
+testing_set = featuresets[700:]
```
Comment (Javiercerna) [inline sentiment_mining.py:30]: Similar as before: Is this related to the 1000 that was referenced before? In any case you could have another constant like `TRAINING_SET_SIZE` if these two are completely independent from one another

**Review by Javiercerna — COMMENTED** [review]
Your code looks easy to read and you're not using any cryptic variable names, I really appreciate that! I of course haven't had the time to download and test your code, but it looks interesting, for sure.

Here are some questions that I got after checking your code:
- Are you familiar with OOP concepts? It's not very clear here since most of your scripts don't benefit much from using classes (implementing them would be a bit overkill), so it's a more general question.
- Do you have a main script which you use to run these small scripts? Or some sort of workflow? Which one do you run first, second, etc. Otherwise it's not very clear to me how you should run your code. Of course a README would help tremendously.

File: bagOfWords.py (hunk 1 above)
Comment (nandakishor123) [reply bagOfWords.py:6]: I will keep that in mind while I code.

File: bagOfWords.py (hunk 2 above)
Comment (nandakishor123) [reply bagOfWords.py:13]: Yes!😅

File: bagOfWords.py (hunk 3 above)
Comment (nandakishor123) [reply bagOfWords.py:28]: Here, the "reviews. csv" was a collection of hotel reviews with each reviews in a line and there were 1600 rows in the file. So I was pulling out the reviews and making another file and that is the code which was serving that purpose.

File: bagOfWords.py (hunk 4 above)
Comment (nandakishor123) [reply bagOfWords.py:30]: It could have been avoided, but I never was much careful back then. This coding happened in 2018, where I was not following much of the programming conventions.

File: bigramProbability.py (hunk 7 above)
Comment (nandakishor123) [reply bigramProbability.py:25]: Totally agree. I had a problem with the conventions back then. Now I think I am fine.

File: feature_extraction.py (hunk 10 above)
Comment (nandakishor123) [reply feature_extraction.py:37]: I did not know that. Will keep in mind.

File: feature_vector.py (hunk 13 above)
Comment (nandakishor123) [reply feature_vector.py:65]: The method is transforming the text data into numerical data, but apparently I never gave a return value for the method. It was there, but I never committed that code into the repository. 😕

File: sentiment_mining.py (hunk 16 above)
Comment (nandakishor123) [reply sentiment_mining.py:18]: Yes would agree with that.

File: sentiment_mining.py (hunk 18 above)
Comment (nandakishor123) [reply sentiment_mining.py:30]: Yes this 700 actually comes from the 1000 which is mentioned above.

**Comment by nandakishor123** [comment]
Thanks for the time you have taken to go through my code. This is some scattered pieces of the code which I had written during my project time. Apparently I do not have a finished version or as you asked the main script from where my pipeline would be running. This coding took place when I was not quite exposed to the standards or conventions I have to be careful of while coding, but I guess I have come far from that stage.
